README.md
# Rest-in-Lease
Sample REST-Service written in Clojure and implemented on AWS Lamda and AWS API Gateway (Dez 2020).
## What it does
The Lambda service is written in Clojure and it takes 6 parameters as inputs in the GET request, such as 
```
https://3xhwkqfrmd.execute-api.us-east-2.amazonaws.com/Test/calc-profit?numLeasePeriods=36&leaseRate=400&deprecRateYrly=0.15&purchasePrice=30000&refiInterestRateYrly=0.03&refiAnnuity=1000
```
It returns an JSON with the results, e.g.
```
{
    "remaining_value": 19075,
    "amortization_periods": 32,
    "contract_profit": 2251,
    "paid_amortization": 30000,
    "paid_interest": 1224
}
```
The service can be called via URL entered in a web browser, via Postmann or from bash: 
```
$ curl --location --request GET 'https://3xhwkqfrmd.execute-api.us-east-2.amazonaws.com/Test/calc-profit?numLeasePeriods=36&leaseRate=400&deprecRateYrly=0.15&purchasePrice=30000&refiInterestRateYrly=0.03&refiAnnuity=1000'
```

Also, there is a simple webpage index.html in the clojure project's src folder. When called from a local webserver (e.g. MiniWebserver etc.) inputs can be entered and results from the Rest-in-Lease Lambda service are displayed.

## How it is implememted
This little sample application consists of the following elements: 
* the AWS Lambda service
* the AWS API Gateway configuration
* (optional) AWS CloudWatch
* (optional) the webpage to demo it 

### AWS Lambda service
The service itself is written in Clojure. 
Special remarks:
* once compiled into a jar-file with class-files, the JVM sees a module.class "rest_in_lease_api_aws.core" with a static method "handler" which is be called by AWS Gateway (see below)
* the Clojure service does not make use of Clojure's popular "ring" http-handling-framework. There was just no need at the end. The job could be done with closure.data.json.
* the current implemented version processes a GET request that ontains the inputs in the query string. The version that processes POST requests with JSON as input is commented out. 
* the request passed to "handler" is a Clojure map with the query-string key-value-pairs as strings. 
* the service creates a Clojure map and returns it to the calling API Gateway. The conversion from the JSON message from API Gateway into a Clojure map and visa versa  is done magically automatically. When using the POST method (code is commented out at the moment), the "body" in the input JSON from API Gateway is a string, containing the actual JSON with the leasing input parameters. This string needs to be converted into a Clojure map explicitely by json/read-string.   

Build:
```
project-folder$ lein uberjar
```
Deploy:
* Logon to: https://aws.amazon.com/de/
* Go to: MyAccount / AWS Management Console / Lambda
* Delete the AWS Lambda service
* Then deploy the uberjar to AWS Lambda:
```
project-folder$ aws lambda create-function   --function-name rest-in-lease-api-aws   --handler rest_in_lease_api_aws.core::handler   --runtime java8   --memory 512   --timeout 10    --zip-file fileb://./target/uberjar/rest-in-lease-api-aws-0.1.0-SNAPSHOT-standalone.jar  --role arn:aws:iam::545854326725:role/service-role/aws-lambda-03-recv-ret-json-2-log-role-cbh95yyk
```
* Check if upload was successful by refreshing the Lambda functions list in the AWS Lambda console
* Then continue with the steps in AWS Gateway, see below.

## AWS API Gateway
* if none yet existent, create a new API Gateway: 
  * protocol = http
  * Deploy / Stages: create a "Test" stage with Automatic Deployment disabled
  * Develop / Routes: set REST end-point path = /calc-profit, add GET route with an in Integration to a Lambda service, select the Lambda service "rest-in-lease-api-aws", set Payload Format Version to 2.0 and Grant API Gateway permission to invoke your Lambda function.
  * CORS: Access-Control-Allow-Origin = * , Access-Control-Allow-Headers = * , Access-Control-Allow-Methods = *
  * Logging: to AWS CloudWatch can be set up, e.g. with log desination "arn:aws:logs:us-east-2:545854326725:log-group:rest-in-lease-aws-api" 
 * When new version of the Lambda function was deployed:
   * Integrations: Manage Integrations: edit / save (to bind the newly deployed Lambda function to this API route) and deploy to stage "Test".
 
 ## AWS CloudWatch
 * For Lambda functions AWS logs into CloudWatch by default, see also:  Lambda > Functions > rest-in-lease-api-aws > Edit monitoring tools
 * For API Gateways logging has to be activated explicitely, see above.
 * To check logs: look under Log Groups for the ones setup for the AWS API Gateway and the AWS Lambda function 

## The webpage
* Run a local webserver on your machine
* Call the index.html page (in src/rest-in-lease-aws-api) 
* Enter input values, press button "Deckungbeitrag Vetrag"
* the results should show up as the first row in the table below.
* Be patient at the first request to the API as AWS Lambda takes a while to start the service. Subsequent calls are responded quickly.
Remark:
* CORS is enabled in the GET request as the request to the AWS API goes to another domain then the domain the webpage came from (i.e. localhost). It is important that the CORS parameters are set in the API Gateway as described above. API Gateway will return the right http headers, e.g. Access-Control-Allow-Origin=* , indicating that API requests from a webpage loaded into the browser from localhost or from any other domain are OK. 





CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-12-05
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-12-05
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rest-in-lease-api-aws/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rest-in-lease-api-aws/compare/0.1.0...0.1.1


test/rest_in_lease_api_aws/core_test.clj
(ns rest-in-lease-api-aws.core-test
  (:require [clojure.test :refer :all]
            [rest-in-lease-api-aws.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/rest_in_lease_api_aws/core.clj

;POST https://3xhwkqfrmd.execute-api.us-east-2.amazonaws.com/Test/calc-profit
;{
;    "numLeasePeriods": 36,
;    "leaseRate" : 550,
;    "deprecRateYrly": 0.15,
;    "purchasePrice": 30000,
;    "refiInterestRateYrly": 0.03,
;    "refiAnnuity": 1000
;}

;GET https://3xhwkqfrmd.execute-api.us-east-2.amazonaws.com/Test/calc-profit?numLeasePeriods=36&leaseRate=650&deprecRateYrly=0.15&purchasePrice=30000&refiInterestRateYrly=0.03&refiAnnuity=1000

(ns rest-in-lease-api-aws.core
  (:require
    [clojure.data.json :as json]
   )
  (:gen-class
   :methods [^:static [handler [Object] Object]]
  )
)

;; ---- helper functions ----
(defn sumUpList [l & {:keys [vecPos] :or {vecPos eval}}] (reduce + (map #(vecPos %) l)))

(defn PadListRight [listToBePadded elem num]
    (defn createExtension [elem num] (if (<= num 0) nil (list* elem (createExtension elem (dec num)))))
    (concat listToBePadded (createExtension elem num))        
)

;; ---- leasing core  functions ----
(defn DeprecCashFlowListForPeriods [deprecRate remLeasePeriods remValue]
    (if (zero? remLeasePeriods) 
        nil 
        (let [deprecOfPeriod (* remValue deprecRate)]
            (list* deprecOfPeriod (DeprecCashFlowListForPeriods deprecRate (dec remLeasePeriods) (- remValue deprecOfPeriod)))
        )
    )
)

(defn AmortAndInterestCashFlowList [intRate annuity remDebt] ; returns a list of tuples ([amort interest] ...)
    (let [interest (* remDebt intRate)
          amort (min remDebt (- annuity interest))
         ]
        (if (>= interest annuity) ; will never end: refactor to exception raising 
            nil ; abort 
            (if (> remDebt 0) 
                (list* [amort interest] (AmortAndInterestCashFlowList intRate annuity (- remDebt amort)))
                nil
            )
        )
     )
)
(defn NetCashFlowList [leaseRate period remValue a&iCfList]
    (if (= (count a&iCfList) 0)
        nil
        (list* 
              (- 
                  (cond
                    (pos? period)   leaseRate
                    (zero? period)  remValue ; sell the asset
                    :else           0
                  )
                  (first (first a&iCfList)) 
                  (last  (first a&iCfList))
              )
              (NetCashFlowList leaseRate (dec period) remValue (rest a&iCfList))
        )
    )
)

(defn contractProfit [numLeasePeriods leaseRate deprecRate purchasePrice refiInterestRate refiAnnuity]
    (def deprecCashFlowList (DeprecCashFlowListForPeriods deprecRate numLeasePeriods purchasePrice))
    (def remValue (- purchasePrice (sumUpList deprecCashFlowList)))
    (def amortAndInterestCashFlowList (AmortAndInterestCashFlowList refiInterestRate refiAnnuity purchasePrice))  ; => ([amort interest] ...)
    (def amortAndInterestCashFlowListExt (PadListRight amortAndInterestCashFlowList [0 0] (+ (- numLeasePeriods (count amortAndInterestCashFlowList)) 1)))

    (def netCashFlowList (NetCashFlowList leaseRate numLeasePeriods remValue amortAndInterestCashFlowListExt))    
    
    (hash-map   
                "paid_amortization"       (Math/round (sumUpList amortAndInterestCashFlowList :vecPos first))
                "paid_interest"           (Math/round (sumUpList amortAndInterestCashFlowList :vecPos last))
                "amortization_periods"    (count amortAndInterestCashFlowList)
                "remaining_value"         (Math/round remValue) 
                "contract_profit"         (Math/round (sumUpList netCashFlowList)) 
    )
)
  
;; ---- funtions that handle interaction with the API Gateway --- 

(defn create-response [input-map]
  (let [  numLeasePeriods      (:numLeasePeriods       input-map) 
          leaseRate            (:leaseRate             input-map)
          deprecRate        (/ (:deprecRateYrly        input-map) 12)
          purchasePrice        (:purchasePrice         input-map)
          refiInterestRate  (/ (:refiInterestRateYrly  input-map) 12)
          refiAnnuity          (:refiAnnuity           input-map)
          return-map           (contractProfit numLeasePeriods leaseRate deprecRate purchasePrice refiInterestRate refiAnnuity)
       ]
    (prn "Inside create-response: return-map=" return-map) 
    return-map
  )
)

;(use '[ring.middleware.json :only [wrap-json-body wrap-json-response]] 
;     '[ring.util.response :only [response]])

(defn -handler [request] ; request is a Clojure map with keys and values in string form; the key "body" contains a JSON document as string
; POST handling
;  (def request-body-string (get-in request ["body"]))
;  (def request-body-as-map (json/read-str request-body-string :key-fn keyword)) ; convert the body string nto a proper Coljure map
;  (prn "handler: request/body as map =" request-body-as-map)
;
;  (create-response request-body-as-map)
;  ;((wrap-json-response create-response) request-body-as-map)

; GET handling
  (def request-params-map-strings (get-in request ["queryStringParameters"])) ; => map with key value pairs, both strings
; (prn "-handler: request-params-map-strings= " request-params-map-strings)
  (def request-params-map-keyed (into {} (for [[k v] request-params-map-strings] [(keyword k) (read-string v) ])))
; (prn "-handler: request-params-map-keyed= " request-params-map-keyed)
  (create-response request-params-map-keyed)
)